overlord/snapshotstate/snapshotstate.go

```diff
@@ -0,0 +1,263 @@
+		if snapst.Active {
+			names = append(names, name)
+		}
+	}
+
+	sort.Strings(names)
+
+	return names, nil
+}
+
+func snapNamesInSnapshotSet(setID uint64, requested []string) (snapsFound, filenames []string, err error) {
+	sort.Strings(requested)
+	found := false
+	err = backendIter(context.TODO(), func(r *backend.Reader) error {
+		if r.SetID == setID {
+			found = true
+			if len(requested) == 0 || strutil.SortedListContains(requested, r.Snap) {
+				snapsFound = append(snapsFound, r.Snap)
+				filenames = append(filenames, r.Name())
+			}
+		}
+
+		return nil
+	})
+	if err != nil {
+		return nil, nil, err
+	}
+	if !found {
+		return nil, nil, client.ErrSnapshotSetNotFound
+	}
+	if len(snapsFound) == 0 {
+		return nil, nil, client.ErrSnapshotSnapsNotFound
+	}
+
+	return snapsFound, filenames, nil
+}
+
+func taskGetErrMsg(task *state.Task, err error, what string) error {
+	if err == state.ErrNoState {
+		return fmt.Errorf("internal error: task %s (%s) is missing %s information", task.ID(), task.Kind(), what)
+	}
+	return fmt.Errorf("internal error: retrieving %s information from task %s (%s): %v", what, task.ID(), task.Kind(), err)
+}
+
+// checkSnapshotTaskConflict checks whether there's an in-progress task for snapshots with the given set id.
+func checkSnapshotTaskConflict(st *state.State, setID uint64, conflictingKinds ...string) error {
+	for _, task := range st.Tasks() {
+		if task.Change().Status().Ready() {
+			continue
+		}
+		if !strutil.ListContains(conflictingKinds, task.Kind()) {
+			continue
+		}
+
+		var snapshot snapshotSetup
+		if err := task.Get("snapshot-setup", &snapshot); err != nil {
+			return taskGetErrMsg(task, err, "snapshot")
+		}
+
+		if snapshot.SetID == setID {
+			return fmt.Errorf("cannot operate on snapshot set #%d while change %q is in progress", setID, task.Change().ID())
+		}
+	}
+
+	return nil
+}
+
+// List valid snapshots.
+// Note that the state must be locked by the caller.
+var List = backend.List
+
+// Save creates a taskset for taking snapshots of snaps' data.
+// Note that the state must be locked by the caller.
+func Save(st *state.State, snapNames []string, users []string) (setID uint64, snapsSaved []string, ts *state.TaskSet, err error) {
+	if len(snapNames) == 0 {
+		snapNames, err = allActiveSnapNames(st)
+		if err != nil {
+			return 0, nil, nil, err
+		}
+	}
+
+	// Make sure we do not snapshot if anything like install/remove/refresh is in progress
+	if err := snapstateCheckChangeConflictMany(st, snapNames, ""); err != nil {
+		return 0, nil, nil, err
+	}
+
+	setID, err = newSnapshotSetID(st)
+	if err != nil {
+		return 0, nil, nil, err
+	}
+
+	ts = state.NewTaskSet()
+
+	for _, name := range snapNames {
+		desc := fmt.Sprintf("Save data of snap %q in snapshot set #%d", name, setID)
+		task := st.NewTask("save-snapshot", desc)
+		snapshot := snapshotSetup{
+			SetID: setID,
+			Snap:  name,
+			Users: users,
+		}
+		task.Set("snapshot-setup", &snapshot)
+		ts.AddTask(task)
```

Comment:
maybe a comment and a check/test that if one these fails we undo all of them? I suppose that's the intention here given they share the same setID